cache_flat_mutation_reader.hh

```diff
@@ -730,20 +749,92 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
         }
     }
 
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
     if (_next_row_in_range) {
-        if (_next_row.range_tombstone_for_row() != _current_tombstone) [[unlikely]] {
-            auto tomb = _next_row.range_tombstone_for_row();
-            auto new_lower_bound = position_in_partition::before_key(_next_row.position());
-            clogger.trace("csm {}: rtc({}, {})", fmt::ptr(this), new_lower_bound, tomb);
-            push_mutation_fragment(mutation_fragment_v2(*_schema, _permit, range_tombstone_change(new_lower_bound, tomb)));
-            _lower_bound = std::move(new_lower_bound);
-            _current_tombstone = tomb;
-            _read_context.cache()._tracker.on_range_tombstone_read();
+        bool remove_row = false;
+
+        auto tomb_expired = [&](tombstone tomb) {
+            return (tomb && tomb.deletion_time < _gc_before);
+        };
+
+        if (_read_context.tombstone_gc_state() // do not compact rows when tombstone_gc_state is not set (used in some unit tests)
+            && _snp->at_latest_version()
+            && _snp->at_oldest_version()) {
+            if (!_next_row.dummy()) {
+                deletable_row& row = _next_row.latest_row();
+                tombstone range_tomb = _next_row.range_tombstone_for_row();
+                auto t = row.deleted_at();
+                t.apply(range_tomb);
+
+                auto row_tomb_expired = [&](row_tombstone tomb) {
+                    return (tomb && tomb.max_deletion_time() < _gc_before);
+                };
+
+                auto is_row_dead = [&](const deletable_row& row) {
+                    return (!row.marker().is_missing() && row.marker().is_dead(_read_time));
+                };
+
+                if (row_tomb_expired(t) || is_row_dead(row)) {
+                    can_gc_fn always_gc = [&](tombstone) { return true; };
+                    row_marker marker = row.marker();
+
+                    try {
+                        with_allocator(_snp->region().allocator(), [&] {
+                            row.compact_and_expire(*_schema, range_tomb, _read_time, always_gc, _gc_before, nullptr);
+                        });
+                        remove_row = row.empty();
+                    } catch (...) {
+                        row.marker() = marker;
+                        throw;
+                    }
+
+                    auto latests_range_tomb = _next_row.get_iterator_in_latest_version()->range_tombstone();
+                    if (tomb_expired(latests_range_tomb)) {
+                        _next_row.get_iterator_in_latest_version()->set_range_tombstone({});
+                    }
+                }
+            } else { // dummy row
+                tombstone range_tomb = _next_row.range_tombstone_for_row();
+                if (tomb_expired(range_tomb)) {
+                    _next_row.ensure_entry_in_latest().row.set_continuous(true);
```

Comment:
Good question. Making these dummy rows continuous is the only way I found to prevent "resurrection" of compacted rows (covered by range tombstone) from underlying source during next read. The test_compact_range_tombstones_on_read test shows it. I suspect this may be not the best solution, probably it's better to compact such dummy rows?

Reply:
Cache populator should compact data from sstables before feeding it to cache since ce96b472d39a6c72f99176c8cf84e171143ae046.